fl4health/clients/ditto_mkmmd_client.py

```diff
@@ -0,0 +1,483 @@
+
+        Args:
+            preds (Dict[str, torch.Tensor]): Prediction(s) of the model(s) indexed by name.
+                All predictions included in dictionary will be used to compute metrics.
+            features: (Dict[str, torch.Tensor]): Feature(s) of the model(s) indexed by name.
+            target: (torch.Tensor): Ground truth data to evaluate predictions against.
+
+        Returns:
+            TrainingLosses: an instance of TrainingLosses containing backward loss and
+                additional losses indexed by name. Additional losses includes each loss component and the global model
+                loss tensor.
+        """
+        # Check that both models are in training mode
+        assert self.global_model.training and self.model.training
+
+        total_loss, additional_losses = self.compute_loss_and_additional_losses(preds, features, target)
+        assert additional_losses is not None
+
+        # Compute ditto drift loss
+        ditto_local_loss = self.get_ditto_drift_loss()
+        additional_losses["ditto_loss"] = ditto_local_loss
+
+        return TrainingLosses(backward=total_loss + ditto_local_loss, additional_losses=additional_losses)
+
+    def validate(self) -> Tuple[float, Dict[str, Scalar]]:
+        """
+        Validate the current model on the entire validation dataset.
+
+        Returns:
+            Tuple[float, Dict[str, Scalar]]: The validation loss and a dictionary of metrics from validation.
+        """
+        # Set the global model to evaluate mode
+        self.global_model.eval()
+        return super().validate()
+
+    def compute_evaluation_loss(
+        self,
+        preds: Dict[str, torch.Tensor],
+        features: Dict[str, torch.Tensor],
+        target: torch.Tensor,
+    ) -> EvaluationLosses:
+        """
+        Computes evaluation loss given predictions (and potentially features) of the model and ground truth data.
+        For Ditto, we use the vanilla loss for the local model in checkpointing. However, during validation we also
+        compute the global model vanilla loss.
+        We also include a sanity check log which computes the ditto drift loss during evaluation to ensure that it
+        is non-zero.
+
+        Args:
+            preds (Dict[str, torch.Tensor]): Prediction(s) of the model(s) indexed by name. Anything stored
+                in preds will be used to compute metrics.
+            features: (Dict[str, torch.Tensor]): Feature(s) of the model(s) indexed by name.
+            target: (torch.Tensor): Ground truth data to evaluate predictions against.
+
+        Returns:
+            EvaluationLosses: an instance of EvaluationLosses containing checkpoint loss and additional losses
+                indexed by name.
+        """
+        # Check that both models are in eval mode
+        assert not self.global_model.training and not self.model.training
+
+        _, additional_losses = self.compute_loss_and_additional_losses(preds, features, target)
```

Comment:
Similar to the train component, we can inherit from Ditto and make this properly inherited.